zebra-consensus/src/transaction.rs

```diff
@@ -153,175 +161,243 @@ where
-                                tracing::warn!(%binding_sig_error, "ignoring");
-                                metrics::counter!("zebra.error.sapling.binding",
-                                                  1,
-                                                  "kind" => binding_sig_error);
-                            });
-                            // Ignore errors until binding signatures are fixed
-                            //.map_err(|e| BoxError::from(Box::new(e)))?;
-                        }
-
-                        let _rsp = redjubjub_verifier
-                            .ready_and()
-                            .await?
-                            .call((bvk, sapling_shielded_data.binding_sig, &shielded_sighash).into())
-                            .boxed();
-
-                        // TODO: stop ignoring binding signature errors - #1939
-                        // async_checks.push(rsp);
-                    }
-
-                    // Finally, wait for all asynchronous checks to complete
-                    // successfully, or fail verification if they error.
-                    while let Some(check) = async_checks.next().await {
-                        tracing::trace!(?check, remaining = async_checks.len());
-                        check?;
-                    }
-
-                    Ok(tx.hash())
-                }
-                Transaction::V5 { .. } => {
-                    unimplemented!("v5 transaction validation as specified in ZIP-216, ZIP-224, ZIP-225, and ZIP-244")
+                    Self::verify_v4_transaction(
+                        &req,
+                        network,
+                        script_verifier,
+                        inputs,
+                        joinsplit_data,
+                        sapling_shielded_data,
+                    )
+                    .await
                 }
+                Transaction::V5 { .. } => Self::verify_v5_transaction(&req, network).await,
             }
         }
         .instrument(span)
         .boxed()
     }
 }
+
+impl<ZS> Verifier<ZS>
+where
+    ZS: Service<zs::Request, Response = zs::Response, Error = BoxError> + Send + Clone + 'static,
+    ZS::Future: Send + 'static,
+{
+    async fn verify_v4_transaction(
+        request: &Request,
+        network: Network,
+        mut script_verifier: script::Verifier<ZS>,
+        inputs: &[transparent::Input],
+        joinsplit_data: &Option<transaction::JoinSplitData<Groth16Proof>>,
+        sapling_shielded_data: &Option<sapling::ShieldedData<sapling::PerSpendAnchor>>,
+    ) -> Result<transaction::Hash, TransactionError> {
+        let mut spend_verifier = primitives::groth16::SPEND_VERIFIER.clone();
+        let mut output_verifier = primitives::groth16::OUTPUT_VERIFIER.clone();
+
+        let mut ed25519_verifier = primitives::ed25519::VERIFIER.clone();
+        let mut redjubjub_verifier = primitives::redjubjub::VERIFIER.clone();
```

Comment:
This is a good move - it lets us only clone the verifiers we need for each transaction version.

(And in the next refactor we can move these verifiers into the per-pool methods.)